plugins/catalog-backend-module-puppetdb/src/providers/PuppetDbEntityProvider.ts

```diff
@@ -0,0 +1,236 @@
+   * @param transformer - A {@link ResourceTransformer} function.
+   *
+   * @private
+   */
+  private constructor(
+    config: PuppetDbEntityProviderConfig,
+    logger: Logger,
+    taskRunner: TaskRunner,
+    transformer: ResourceTransformer,
+  ) {
+    this.config = config;
+    this.logger = logger.child({
+      target: this.getProviderName(),
+    });
+    this.scheduleFn = this.createScheduleFn(taskRunner);
+    this.transformer = transformer;
+  }
+
+  /** {@inheritdoc @backstage/plugin-catalog-backend#EntityProvider.connect} */
+  async connect(connection: EntityProviderConnection): Promise<void> {
+    this.connection = connection;
+    await this.scheduleFn();
+  }
+
+  /** {@inheritdoc @backstage/plugin-catalog-backend#EntityProvider.getProviderName} */
+  getProviderName(): string {
+    return `puppetdb-provider:${this.config.id}`;
+  }
+
+  /**
+   * Creates a function that can be used to schedule a refresh of the catalog.
+   *
+   * @param taskRunner - The instance of {@link TaskRunner}.
+   *
+   * @private
+   */
+  private createScheduleFn(taskRunner: TaskRunner): () => Promise<void> {
+    return async () => {
+      const taskId = `${this.getProviderName()}:refresh`;
+      return taskRunner.run({
+        id: taskId,
+        fn: async () => {
+          const logger = this.logger.child({
+            class: PuppetDbEntityProvider.prototype.constructor.name,
+            taskId,
+            taskInstanceId: uuid.v4(),
+          });
+          try {
+            await this.refresh(logger);
+          } catch (error) {
+            logger.error(`${this.getProviderName()} refresh failed`, error);
+          }
+        },
+      });
+    };
+  }
+
+  /**
+   * Refreshes the catalog by reading nodes from PuppetDB and registering them as Resource Entities.
+   *
+   * @param logger - The instance of a Logger.
+   */
+  async refresh(logger: Logger) {
+    if (!this.connection) {
+      throw new Error('Not initialized');
+    }
+
+    const { markReadComplete } = trackProgress(logger);
+    const entities = await readPuppetNodes(this.config, {
+      logger,
+      transformer: this.transformer,
+    });
+    const { markCommitComplete } = markReadComplete(entities);
+
+    await this.connection.applyMutation({
+      type: 'full',
+      entities: [...entities].map(entity => ({
+        locationKey: this.getProviderName(),
+        entity: withLocations(this.config.host, entity),
+      })),
+    });
+    markCommitComplete(entities);
+  }
+}
+
+/**
+ * Ensures the entities have required annotation data.
+ *
+ * @param host - The host of the PuppetDB instance.
+ * @param entity - The entity to add the annotations to.
+ *
+ * @returns Entity with @{@link ANNOTATION_LOCATION} and @{@link ANNOTATION_ORIGIN_LOCATION} annotations.
+ */
+function withLocations(host: string, entity: Entity): Entity {
+  const location = new URL(host);
+  location.pathname = `${ENDPOINT_NODES}/${entity.metadata?.name}`;
```

Comment:
Can the puppetdb ever be configured on a subpath? Could you have it on `https://infra.mycompany.net/puppetdb`?

Maybe you should append the path to the base instead (ensuring that they get joined with a single slash, whether the app-config accidentally had an extra trailing slash in it or not; maybe the config reader can strip out those already at the source)

Reply:
We already trim the extra trailing slash as per https://github.com/backstage/backstage/pull/16184/files#diff-0ce4d0baa04b46c5b299bdd09d351d94c160e92bfa54a3a76c1e3072cc900180R86